### PR TITLE
ADR-8 Marconi Testing Strategy

### PR DESCRIPTION
**Pull request by james-iohk**
Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense and have useful messages
    - [ ] Important changes are reflected in changelog.d of the affected packages
    - [ ] Relevant tickets are mentioned in commit messages
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [ ] Targeting main unless this is a cherry-pick backport
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] If relevant, reference the ADR in the PR and reference the PR in the ADR
    - [ ] Reviewer requested

### CODE REVIEW

File: doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  See RPC-1 in https://github.com/input-output-hk/marconi/issues/93 for checklist of test cases.
+
+* We will produce tests for the RPC API methods to cover a board set of response data comparing against a source of truth
+  These tests will compare against results produced by `cardano-db-sync <https://github.com/input-output-hk/cardano-db-sync>`_
+  These tests will ideally be run on mainnet
+  see RPC-2 in https://github.com/input-output-hk/marconi/issues/93 for checklist of test cases.
+
+* We will not have any end-to-end tests that query maroni's db because this does not always store relevant data and it does not resemble how the user should be querying.
+
+* We will produce sync performance tests to be run on each public network: preview, preprod and mainnet. (See ADR 7: Marconi observability for details)
```
Comment (berewt) [inline doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst:79]: do we set threshold that would lead to failing tests result?

File: doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+Argument
+--------
+
+Being the primary customer-facing application, Marconi-Sidechain will undergo regular updates and enhancements to meet evolving customer needs, which exposes the risk of regressions in its functionality and accepted performance.
+
+Consequently, an ongoing process is necessary to validate the compatibility and performance of new features or modifications whilst verifying that legacy features remain working correctly.
+
+Alternative solutions
+---------------------
+When testing marconi-chain-index we can take a similar approach but must also:
```
Comment (berewt) [inline doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst:92]: We definitely have to decide the future of Marconi-chain-index first.

**Review by berewt — COMMENTED** [review]
No blockers but a few remarks and questions.

File: doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+A general approach to functional testing Marconi-Sidechains covering all test levels.
+The scope of this document is `marconi-sidechain`, not other Marconi applications such as `marconi-chain-index`.
+
+Decision
+--------
+
+* We will design and implement tests at the following levels:
+    * Unit tests are run locally on individual functions and modules.
+    * Integration tests can be easilly run locally and on commit using cardano-testnet for a private testnet (soon to use the node emulator).
+    * End-to-end tests are run on a public network with a real cardano node. These can be cumbersome to setup and run due requiring a fully synced cardano-node and funded wallet, and are slow to execute due to block propogation. Therefore, we will only run automated end-to-end tests nightly.
```
[the remark this answers is not in the document]
Comment (james-iohk) [reply doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst]: No, you're right. We won't be submitting any transactions for marconi e2e testing. Thanks.

File: doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+* We will write tests in Haskell where possible so these tests easily integrate into existing infrastructure.
+  Only use other tools if there are strong benefits.
```
[the remark this answers is not in the document]
Comment (james-iohk) [reply doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst:33]: True. If we want to use a specific tool for performance testing that would alleviate us from implementing the framework for performance testing then that is a strong enough benefit.

File: doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  These tests will ideally be run on mainnet because there is more interesting transaction data there and so are more likely to catch an edge-case.
+  These tests should be run either after a sync test or on an instance of the process under test that is always running fully synced.
+  See RPC-1 in https://github.com/input-output-hk/marconi/issues/93 for checklist of test cases.
+
+* We will produce tests for the RPC API methods to cover a board set of response data comparing against a source of truth
+  These tests will compare against results produced by `cardano-db-sync <https://github.com/input-output-hk/cardano-db-sync>`_
+  These tests will ideally be run on mainnet
+  see RPC-2 in https://github.com/input-output-hk/marconi/issues/93 for checklist of test cases.
+
+* We will not have any end-to-end tests that query maroni's db because this does not always store relevant data and it does not resemble how the user should be querying.
```
[the remark this answers is not in the document]
Comment (james-iohk) [reply doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst:77]: Simply an argument for always testing using the RPC interface, for if someone did consider testing by querying the db. Unless you think there is a valid reason to?

File: doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst (hunk 1 above)
Comment (james-iohk) [reply doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst:79]: Yeah, definitely. 
I didn't want to put details on performance testing here but it is worth referencing the other ADR for those looking.

File: doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  Only use other tools if there are strong benefits.
+
+* We will produce test reports for each run which will be hosted publicly in Allure format.
+  In lieu of a Haskell adapter for Allure, `tasty-test-reporter` can be used.
+  It is not important to retain previous reports triggered by nightly or for PR commits runs.
+  Reports produced for a release should be hosted indefinitely.
+
+UNIT TESTING:
+
+* We will test indexers and filters with as many positive and negative combinations as possible using data-driven and property tests.
```
[the remark this answers is not in the document]
Comment (james-iohk) [reply doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst:42]: Sounds useful, even if it is a significant task. I'll add a point but feel free to contribute further detail.

[the remark this answers is not in the document]
Comment (berewt) [reply doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst:42]: It is and it could actually be a transversal useful testing utility on its own, so it might be out of scope here.

File: doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (berewt) [reply doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst:77]: No, I just didn't get that meaning from the original sentence. To me it's much clearer phrased in the positive way of your answer than it is in the ADR.

File: doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (james-iohk) [reply doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst:43]: I added a supporting line on this point. It's ok if it turns out to be useful for more than just marconi testing.

File: doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst (hunk 1 above)
Comment (koslambrou) [reply doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst:81]: I think we need to mention this in this ADR. The other ADR is about identify how to "observe" the behaviour of `marconi-sidechain`.
Maybe put anything related to performance tests wrt to sync time, memory usage, query performance, etc should go here?

**Review by koslambrou — COMMENTED** [review]
In general, seems to go in a nice direction. 

I think syncing and query benchmarking should be outlined here in more detail.

File: doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst (hunk 1 above)
Comment (james-iohk) [reply doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst]: Thanks for the feedback. I have now added some more detail on the sync and query performance testing.

File: doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst (hunk 2 above)
Comment (james-iohk) [reply doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst:110]: Agreed. I wanted to keep this point about testing the flags somewhere.

File: doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+* We will produce test reports for each run which will be hosted publicly in Allure format.
+  In lieu of a Haskell adapter for Allure, `tasty-test-reporter <https://hackage.haskell.org/package/tasty-test-reporter>`_ can be used.
+  It is not important to retain previous reports triggered by nightly or for PR commits runs.
+  Reports produced for a release should be hosted indefinitely.
+
+UNIT TESTING:
+
+* We will test indexers and filters with as many positive and negative combinations as possible using data-driven and property tests.
+  To support this, we will produce a generator to act like a "mockchain" to generate a variety of: blocks, transations and roll forward/backward events.
```
[the remark this answers is not in the document]
Comment (james-iohk) [reply doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst]: Added some points relating to this in unit and integration sections.

File: doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+  .  To support this, we will produce a generator to act like a "mockchain" to generate a variety of: blocks, transations and roll forward/backward events.
+
+* We will only test functions and modules at lower layers to the CLI or RPC and not aim to cover these interfaces because we will get greater coverage of those with the integration tests.
+
+* We will plan for and produce tests alongside implementation of new product features.
+
+INTEGRATION TESTING:
+
+* We will plan test coverage and also aim to produce them alongside implementation of new product features.
+  It is acceptable to have separate tasks to produce the tests but them must be complete within the same PI cycle.
```
Comment (koslambrou) [inline doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst:56]: Typo: but they

File: doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+  This can provide faster feedback in lieu of automation testing being produced.
+  Should not be carried out by the engineer who implemented the feature (due to assumption bias).
+    
+* We will produce tests for each of the RPC API methods to stress the filters
+  These tests will assert against known correct response content.
+  These tests will ideally be run on mainnet because there is more interesting transaction data there and so are more likely to catch an edge-case.
+  These tests should be run either after a sync test or on an instance of the process under test that is always running fully synced.
+  See RPC-1 in https://github.com/input-output-hk/marconi/issues/93 for a checklist of test cases.
+
+* We will produce tests for the RPC API methods to cover a board set of response data comparing against a source of truth
```
Comment (koslambrou) [inline doc/read-the-docs-site/adr/0008-marconi-sidechain-testing-strategy.rst:82]: Forgot a dot at the end of the sentence

**Review by koslambrou — APPROVED** [review]
